lib/puppet/parser/functions/create_component_app.rb

```diff
@@ -0,0 +1,38 @@
+# Creates an instance of an application. Treats 'components' param as special and maps nodes to components.
+#
+# create_component_app($app_name, $instance_name, $params_and_components)
+#
+# create_component_app('wordpress::app', 'my_wordpress',
+#   {'db_password' => 'secret',
+#    'components' => {
+#      'Wordpress_app::database' => ['node1.example.com'],
+#      'Wordpress_app::web' => ['node2.example.com', node3.example.com']}})
+#
+#
+# @param app_name The application to create an instance of
+# @param instance_name The title of to the application instance to declare
+# @param params The params to create the application with. Expects a components hash that maps components to node names.
+module Puppet::Parser::Functions
+  newfunction(:create_component_app, arity: 3) do |args|
+    app_name, instance_name, params = args
+
+    # TODO: Should we also accept nodes?
+    nodes = {}
+    components = params['components']
+    components.each do |comp_name, node_names|
+      node_names.each do |node_name|
+        comp = Puppet::Resource.new(nil, "#{comp_name}[#{app_name}-#{node_name}]", {})
```

Comment:
Seems like we should refer to `instance_name` here instead of or in addition to app_name.